### PR TITLE
Added 'validate_tags' to test tags across middleware pages

### PR DESCRIPTION
**Pull request by jkandasa**
Added new function `validate_tags` in `Validatable` class and implemented across Middleware pages.
* Added support function for `validate_tags`
* Added tests (`test_tags`) for Middleware `Provider`, `Server` and `Deployment` pages
* For `Datasource` page test not included, blocks: https://github.com/ManageIQ/manageiq/issues/9211

{{pytest: cfme/tests/cloud_infra_common/test_events.py -v --long-running --use-provider nested-rhevm34}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit f281d50e48fb6023827ebd76b79e389bdf91ef4f:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/common/__init__.py [hunk 1]
```diff
@@ -79,14 +84,51 @@ def add_tag(self, tag, single_value=False):
...
+        t_cls2 = aliased(cfmedb()['classifications'])
+        t_tgg = aliased(cfmedb()['taggings'])
+        query = cfmedb().session.query(t_cls1.tag_id, t_tgg.taggable_id.label('db_id'),
+                                       t_cls2.description.label('category'),
+                                       t_cls1.description.label('tag_name'), t_cls1.single_value)\
+            .join(t_cls2, t_cls1.parent_id == t_cls2.id)\
+            .join(t_tgg, t_tgg.tag_id == t_cls1.tag_id)\
+            .filter(t_tgg.taggable_id == self.db_id)\
+            .filter(t_tgg.taggable_type == self.taggable_type)
+        return query.all()
```
Comment (psav) [inline cfme/common/__init__.py]: I presume this will give a different output to the "ui" function above it. This could lead to problems if we write a test that could use "either" method.

File: cfme/common/__init__.py [hunk 2]
```diff
@@ -326,3 +368,88 @@ def validate_properties(self):
...
+            self.remove_tags(tags=self._get_tags(db_tags=tags_db_raw))
+            tags_db_raw = self.get_tags(method='db')
+        assert len(tags_db_raw) == 0, "Some of tags still available in database!"
+        self.add_tags(tags)
+        self._validate_tags_internal(reference_tags=tags)
+
+    def _validate_tags_internal(self, tag="My Company Tags", reference_tags=None):
+        """
+        Validation method which check tagging between UI and database
+        To use this method, `seld` should be extended `Taggable` class
```
Comment (psav) [inline cfme/common/__init__.py]: ```seld```??

File: cfme/common/__init__.py (hunk 1 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @psav Yes, I have added to convert to common output by `def _get_tags(self, tag="My Company Tags", ui_tags=None, db_tags=None):` https://github.com/ManageIQ/integration_tests/pull/3080/files#diff-b4ea81c2fa5c5c4440e5a7980ffd22a2R427

What do you think on this? Is this ok or Shall I move everything in to GUI format?

File: cfme/common/__init__.py (hunk 2 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @psav I see blank comment here. Could you please type once again?

File: cfme/common/__init__.py (hunk 1 above)
Comment (psav) [inline cfme/common/__init__.py]: Right - I'm unsure if the way forward will change - ultimately we will want it to return "Tag" objects, so that they are usable elsewhere. We have not done this in the past, and it will be something new, I would say, if you can lets make it collapse to the GUI format for compatibility purposes, but that in the future we are already thinking about how to return objects, possibly by querying the db to help instantiate with the correct information.

File: cfme/common/__init__.py (hunk 2 above)
Comment (mfalesni) [inline cfme/common/__init__.py]: ```To use this method, `seld` should be extended `Taggable` class```
seld vs self. But anyway this sentence does not make much sense, I would reword it.

File: cfme/common/__init__.py [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 from cached_property import cached_property
 from cfme.fixtures import pytest_selenium as sel
 from cfme.web_ui import CheckboxTree, flash, form_buttons, mixins, toolbar
+from sqlalchemy.orm import aliased
 from utils import attributize_string, version
+from utils.db import cfmedb
+from utils.varmeth import variable
 
 pol_btn = partial(toolbar.select, "Policy")
+Tag = namedtuple('Tag', 'category, tag_name single_value')
```
Comment (mfalesni) [inline cfme/common/__init__.py]: I guess we could use the Tag classes from Configuration directly.

File: cfme/common/__init__.py [hunk 4]
```diff
@@ -79,14 +84,51 @@ def add_tag(self, tag, single_value=False):
...
+
     def remove_tag(self, tag):
         self.load_details(refresh=True)
         mixins.remove_tag(tag)
 
+    def remove_tags(self, tags):
+        """Remove list of tags
+        tags: List of `Tag` namedtuple
+        """
+        for _tag in tags:
```
Comment (mfalesni) [inline cfme/common/__init__.py]: I guess there is no need to prefix the `tag` with `_`

File: cfme/common/__init__.py (hunk 1 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @psav merged `_get_tags` directly with `Taggable` class

**Comment by dajoRH** [comment]
Lint report for commit 3591284cafff14bbc7a95866d122c2c79bbcd08d:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/common/__init__.py (hunk 3 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni I see `Tag` class from `Configuration` has action methods. I could not get how it is working and how `single_value` managed here. Could you please point me some document for this `Tag` class?
https://github.com/ManageIQ/integration_tests/blob/master/cfme/configure/configuration/__init__.py#L1770

File: cfme/common/__init__.py (hunk 4 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni `_tag` used in internal hence if we put `_` may be `private` scope?
I will change this.

File: cfme/common/__init__.py (hunk 2 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni You are right. I have updated. Please give your feedback to make it much better.

File: cfme/common/__init__.py (hunk 4 above)
Comment (mfalesni) [inline cfme/common/__init__.py]: it is an internal variable defined in the scope of the function/method, therefor it will not clash with outer world.

Scoping in python is lexical.

**Comment by dajoRH** [comment]
Lint report for commit 21fc3ae534a3cd64f0090ede4fae8754e7dc0843:

`cfme/tests/middleware/test_middleware_deployment.py`:
- :bangbang: Line 127:57: **E231** *missing whitespace after ','*

Please, rectify these issues :smirk: .
*CFME QE Bot*

File: cfme/common/__init__.py (hunk 3 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni Now using `Tag` and `Category` classes from `Configuration`

File: cfme/common/__init__.py (hunk 4 above)
Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni Now I have changed the line as follows,
```
for _tag in tags:
  self.remove_tag(tag=_tag)
```
If I pass `tag=tag` will give bit confusion, Is that fine?

File: cfme/common/__init__.py (hunk 3 above)
Comment (mfalesni) [inline cfme/common/__init__.py]: The single_value is handled inside https://github.com/ManageIQ/integration_tests/blob/master/cfme/configure/configuration/__init__.py#L1712

Comment (jkandasa) [inline cfme/common/__init__.py]: @mfalesni Yes, thank you! I noticed bit later and updated the code.

**Comment by dajoRH** [comment]
Lint report for commit 3a0cae36ae11a063e97d141f5326cf570d788f2c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b001a3430b402cea2fc79e1bd0703cef4c3ebbfd:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9b6c66e07997c3482382249ebf1f9d78458e9797:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ad712d7b84e3c0a7e218baed31391a5d1bbab5c3:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 35e3a289118e8021766189a49efe6daf3174e8af:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7565947d67fe39b94657d871687077335c5ce67b:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@psav @jkrocil I will need a short input from either of you, @jkandasa also had t ochange the tagging methods usage outside of the middleware scope. Can you just do a quick glance over it so I am not missing anything else?

**Comment by jkandasa** [comment]
@psav @jkrocil Can you please have a look and merge this PR if you do not have any comments?

**Comment by mfalesni** [comment]
I will do a followup on this one to fix things I don't like much :)
